### PR TITLE
Use freedesktop Thumbnailer (Tumbler) to get wallpaper thumbnails

### LINKED ISSUE #8: use Tumbler to get wallpaper thumbs
We should probably be using Tumbler daemon to get the wallpaper thumbs instead of carrying that extra code ourselves.

Launchpad Details: [#LP1196280](https://bugs.launchpad.net/bugs/1196280) Daniel Fore - 2013-06-30 17:06:00 +0000

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/43027898-use-tumbler-to-get-wallpaper-thumbs?utm_campaign=plugin&utm_content=tracker%2F57462545&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F57462545&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by davidmhewitt**
Fixes #8 

This PR slightly reduces the size of the wallpaper previews from 168 pixels wide to 128 pixels wide as the largest thumbnail that Tumbler generates is 256 pixels. This allows us to still display nice 2x previews on HiDPI displays.

### CODE REVIEW

**Comment by davidmhewitt** [comment]
@jeremypw I believe I've addressed all of your issues. Thanks for the review :smile:

File: src/ThumbnailGenerator.vala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    public void dequeue_all () {
+        foreach (var handle in handles) {
+            try {
+                thumbnailer.dequeue (handle);
+            } catch (IOError e) {
+                warning ("Unable to tell thumbnailer to stop creating thumbnails: %s", e.message);
+            }
+        }
+    }
+
```
Comment (jeremypw) [inline src/ThumbnailGenerator.vala]: Shouldn't the handles and delegates arraylist/map be emptied too? Or is the "ready" signal emitted when dequeuing?

File: src/Views/Wallpaper.vala [hunk 2]
```diff
@@ -418,105 +408,77 @@ public class Wallpaper : Gtk.Grid {
...
+                // Skip the default_wallpaper as seen in the description of the
+                // default_link variable
+                if (uri == DEFAULT_LINK) {
+                    continue;
+                }
 
-                    var wallpaper = new WallpaperContainer (uri);
-                    wallpaper_view.add (wallpaper);
-                    wallpaper.show_all ();
+                var wallpaper = new WallpaperContainer (uri, file_info.get_attribute_as_string (FileAttribute.THUMBNAIL_PATH), file_info.get_attribute_boolean (FileAttribute.THUMBNAIL_IS_VALID));
```
Comment (jeremypw) [inline src/Views/Wallpaper.vala]: Line too long (elementary code style hard limit is 120 char)

File: src/Views/Wallpaper.vala [hunk 3]
```diff
@@ -418,105 +408,77 @@ public class Wallpaper : Gtk.Grid {
...
         try {
-            // Count the # of wallpapers
-            int count = IOHelper.count_wallpapers(directory);
-            if (count == 0) {
-                folder_combo.set_sensitive (true);
-            }
-
             // Enumerator object that will let us read through the wallpapers asynchronously
-            var e = yield directory.enumerate_children_async (FileAttribute.STANDARD_NAME + "," + FileAttribute.STANDARD_TYPE + "," + FileAttribute.STANDARD_CONTENT_TYPE, 0, Priority.DEFAULT);
+            var e = yield directory.enumerate_children_async (string.joinv (",", required_file_attrs), 0, Priority.DEFAULT);
```
Comment (jeremypw) [inline src/Views/Wallpaper.vala]: Line too long (elementary code style hard limit is 120 char)

File: src/Views/Wallpaper.vala [hunk 4]
```diff
@@ -620,7 +580,7 @@ public class Wallpaper : Gtk.Grid {
                 }
 
                 // Add the wallpaper name and thumbnail to the IconView
-                var wallpaper = new WallpaperContainer (local_uri);
+                var wallpaper = new WallpaperContainer (local_uri, info.get_attribute_as_string (FileAttribute.THUMBNAIL_PATH), info.get_attribute_boolean (FileAttribute.THUMBNAIL_IS_VALID));
```
Comment (jeremypw) [inline src/Views/Wallpaper.vala]: Line too long (elementary code style hard limit is 120 char)

**Comment by davidmhewitt** [comment]
@jeremypw Issues addressed, thanks again!

File: src/Views/Wallpaper.vala [hunk 5]
```diff
@@ -418,105 +408,80 @@ public class Wallpaper : Gtk.Grid {
         }
     }
 
-    private async void load_wallpapers (string basefolder, Cancellable cancellable) {
+    private async void load_wallpapers (string basefolder, Cancellable cancellable, bool toplevel_folder = true) {
         if (cancellable.is_cancelled () == true) {
```
Comment (jeremypw) [inline src/Views/Wallpaper.vala]: == true is redundant,  is_cancelled () already returns a boolean.

File: src/Views/Wallpaper.vala [hunk 6]
```diff
@@ -418,105 +408,80 @@ public class Wallpaper : Gtk.Grid {
...
             // Enumerator object that will let us read through the wallpapers asynchronously
-            var e = yield directory.enumerate_children_async (FileAttribute.STANDARD_NAME + "," + FileAttribute.STANDARD_TYPE + "," + FileAttribute.STANDARD_CONTENT_TYPE, 0, Priority.DEFAULT);
+            var attrs = string.joinv (",", required_file_attrs);
+            var e = yield directory.enumerate_children_async (attrs, 0, Priority.DEFAULT);
+            FileInfo file_info;
 
-            while (true) {
+            // Loop through and add each wallpaper in the batch
+            while ((file_info = e.next_file ()) != null) {
                 if (cancellable.is_cancelled () == true) {
```
Comment (jeremypw) [inline src/Views/Wallpaper.vala]: == true is redundant, is_cancelled () already returns a boolean.

**Pull request merged**